### ISSUE TITLE
A few a11y improvements

### ISSUE THREAD

**Issue by simonleadbetter** [verbb/formie#529]
Hi Verrb

We are using Formie more often in our projects. It’s a great plugin.

As a company, we prioritise a11y and try to do everything we can to continually improve our clients’ websites. Whilst checking our recent site build, we’ve stumbled across a few minor accessibility issues with Formie. I thought we’d share them and ask that you consider making these updates for a future release.

There are three in total…

Attribute :aria-checked not allowed on element

A hidden field is injected alongside any checkbox field. 

```<input type="hidden" aria-checked="false" aria-hidden="true" name="fields[termsAndConditions]" value=""><div class=“fui-checkbox">```

As the field type is hidden, it can not contain the aria-checked attribute and it should be removed.

Attribute placeholder is only allowed when the input type is text, search, url, tel, email, password or number.

```<input  type="checkbox" id="fields-contact-us-terms-and-conditions" class="fui-input fui-checkbox-input" name="fields[termsAndConditions]" value="1" placeholder="" required data-fui-message="" aria-required="true" aria-checked="false" />```

On the checkbox field, the placeholder attribute should be removed.

Attribute aria-required is unnecessary for elements that have attribute required

Finally, anywhere the required attribute is added you can remove the aria-required attribute. 

We often get clients running automated tests and this appears on the reports, so, although this does not cause any accessibility issues per se, it isn’t needed as the required attributed is widely supported.

———

I hope this is useful info and an acceptable change request.

Kind regards,

Simon

**Comment by engram-design** [comment]
Thanks for starting the conversation on this, we also take it pretty seriously, so we'll so all we can.

> Attribute :aria-checked not allowed on element
> A hidden field is injected alongside any checkbox field.
> <div class=“fui-checkbox">
> As the field type is hidden, it can not contain the aria-checked attribute and it should be removed.

We can't really remove the hidden field as we rely on this to record when there are "no" values checked. If we omitted a value for the POST param for the field, that's not the same as saying "I want none checked". This is an important distinction when editing submissions, or when editing a multi-page submission (changing a value on a previous page). I think we'll just ensure we don't add the checked attributes to these special hidden fields.

It's worth clarifying if this is an Agree field (single checkbox) or a Checkboxes field.

> Attribute placeholder is only allowed when the input type is text, search, url, tel, email, password or number.
> On the checkbox field, the placeholder attribute should be removed.

Yep, that certainly shouldn't be there. I believe this is only an issue with the Agree field?

> Attribute aria-required is unnecessary for elements that have attribute required
> Finally, anywhere the required attribute is added you can remove the aria-required attribute.
> We often get clients running automated tests and this appears on the reports, so, although this does not cause any accessibility issues per se, it isn’t needed as the required attributed is widely supported.

Hmm, interesting. I'll do some further research here, as I had thought it "best practice" to include both, but that may have changed in recent years. No real harm removing it, just want to ensure it's the right thing to do.

Thanks!

**Comment by simonleadbetter** [comment]
np

Sorry for not being clearer about the first point. I only meant remove the attribute and not the entire field.

**Comment by engram-design** [comment]
Improved in [1.4.13](https://github.com/verbb/formie/releases/tag/1.4.13)
